File: src/main/java/com/gabrielsson/PizzaResource.java
package com.gabrielsson;

import com.gabrielsson.model.Ingredient;
import com.gabrielsson.model.Pizza;
import org.eclipse.microprofile.metrics.MetricRegistry;
import org.eclipse.microprofile.metrics.Tag;
import org.eclipse.microprofile.metrics.annotation.Timed;
import org.paukov.combinatorics3.Generator;

import javax.enterprise.inject.Any;
import javax.inject.Inject;
import javax.ws.rs.Consumes;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@Path("/pizzas")
public class PizzaResource {
    @Inject
    @Any
    CityService cityService;

    @Inject
    MetricRegistry registry;

    public PizzaResource() {

    }

    @POST
    @Timed
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public List<Pizza> pizzas(List<Ingredient> ingredients) {
        registry.counter("pizza.menu.counter").inc();
        ingredients.forEach(i -> registry
                .counter("pizza.menu.ingredient.counter",
                        new Tag("ingredient", i.getName()))
                .inc());

        List<List<String>> pizzas = Generator
                .subset(ingredients.stream()
                        .skip(Math.max(0, ingredients.size() - 16))
                        .map(Ingredient::getName)
                        .collect(Collectors.toList()))
                .simple()
                .stream()
                .collect(Collectors.toList());

        List<String> names = cityService.newPizzaNames(pizzas);
        return IntStream.range(0, names.size())
                .boxed()
                .map(i -> Pizza.builder()
                        .name(names.get(i))
                        .ingredients(pizzas.get(i).stream()
                                .map(s -> Ingredient.builder()
                                        .name(s)
                                        .build())
                                .collect(Collectors.toList()))
                        .build())
                .collect(Collectors.toList());

    }
}

File: src/main/scala/com/gabrielsson/PizzaResource.scala
package com.gabrielsson

import org.eclipse.microprofile.metrics.annotation.Timed
import org.eclipse.microprofile.metrics.{MetricRegistry, Tag}
import org.eclipse.microprofile.rest.client.inject.RestClient
import org.paukov.combinatorics3.Generator

import javax.inject.Inject
import javax.ws.rs.core.MediaType
import javax.ws.rs.{Consumes, POST, Path, Produces}
import scala.collection.JavaConverters.{iterableAsScalaIterableConverter, mapAsJavaMapConverter, seqAsJavaListConverter}
import scala.collection.breakOut

@Path("/pizzas")
class PizzaResource {

  @Inject
  @RestClient
  protected var cityService: CityService = null

  @Inject
  protected var registry: MetricRegistry = _

  @POST
  @Timed
  @Consumes(Array(MediaType.APPLICATION_JSON))
  @Produces(Array(MediaType.APPLICATION_JSON))
  def pizzas(ingredients: Array[String]) = {
    registry.counter("pizza.menu.counter").inc()

    ingredients.foreach(i => registry
      .counter("pizza.menu.ingredient.counter",
        new Tag("ingredient", i))
      .inc())

    val permutations = getPermutations(ingredients)
    val names = cityService.newPizzaNames(permutations)

    (names zip permutations)(breakOut).toMap.asJava
  }

  def getPermutations(ingredients: Array[String]): Array[Array[String]] = {

    val target: java.util.List[String] = ingredients.toSeq.asJava


    val subsets = Generator.subset(target).simple().asScala.toArray


    subsets.map(_.asScala).map(_.toArray)
  }
}
